layabout.py

```diff
@@ -1,3 +1,287 @@
+
+    Inherits from both :obj:`LayaboutError` and the built-in
+    :obj:`ConnectionError` for convenience in exception handling.
+    """
+
+
+class Layabout:
+    """
+    An event handler on top of the Slack RTM API.
+
+    Args:
+        name: A unique name for this :obj:`Layabout` instance.
+        env_var: The environment variable to try to load a Slack API token
+            from.
+
+    Attributes:
+        name (str): A unique name for this :obj:`Layabout` instance.
+        slack (SlackClient): A :obj:`slackclient.SlackClient` instance.
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
+        self.name = name
+        self._env_var = env_var
+        self._token: str = None
+        self.slack: SlackClient = None
+        self._handlers: _Handlers = defaultdict(list)
+
+    def handle(self, type: str, kwargs: dict = None) -> Callable:
+        """
+        Register an event handler with the :obj:`Layabout` instance.
+
+        Args:
+            type: The name of a Slack RTM API event to be handled. As a
+                special case, although it is not a proper RTM event, ``*`` may
+                be provided to handle all events. For more information about
+                available events see the
+                `Slack RTM API <https://api.slack.com/rtm>`_.
+            kwargs: Optional arbitrary keyword arguments passed to the event
+                handler when the event is triggered.
+
+        Returns:
+            A decorator that validates and registers a Layabout event handler.
+
+        Raises:
+            TypeError: If the decorated :obj:`Callable`'s signature does not
+                accept at least 2 parameters.
+        """
+        def decorator(fn: Callable) -> Callable:
+            # Validate that the wrapped callable is a suitable event handler.
+            sig = signature(fn)
+            if len(sig.parameters) < 2:
+                raise TypeError("Layabout event handlers take at least 2 "
+                                "positional arguments, a slack client and an "
+                                "event")
```

Comment:
My brain wants you to include the found parameters so that it could better contextualize this error.

Reply:
Like, you want to include how many parameters were actually found? Or to be checking the names of parameters? :thinking:

Reply:
I think he just wants more context for the error.  I don't think checking the argument names is a good idea, but possibly printing out the found function signature in the error with `str(signature)` would make sense. I would want the handler function name too so it looks like my definition.

Reply:
I swapped this out for
```python
sig = signature(fn)
num_params = len(sig.parameters)
if num_params < 2:
    plural = '' if num_params else 's'
    raise TypeError(f"{fn.__name__}{sig} takes {num_params} "
                    f"positional argument{plural}, but event "
                    "handlers require at least 2")
```
which, for an invalid handler raises something like
```python
>>> @layabout.handle('hello')
... def invalid_handler(slack):
...    pass
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tucker/Projects/layabout/layabout.py", line 105, in decorator
    raise TypeError(f"{fn.__name__}{sig} takes {num_params} "
TypeError: invalid_handler(slack) takes 1 positional argument, but event handlers require at least 2
>>>
```

Given that Python will raise a `TypeError` for a normal function like this
```python
>>> def fn(n):
...     pass
...
>>> fn(1, 2)
Traceback (most recent call last):             
  File "<stdin>", line 1, in <module>          
TypeError: fn() takes 1 positional argument but 2 were given
>>>
```
I'm pretty satisfied with this change. Thanks for the suggestions, @RadicalZephyr and @rawrgulmuffins!

Reply:
```python
@staticmethod
def _format_parameter_error_message(name: str, sig: Signature,
                                    num_params: int) -> str:
    if num_params == 0:
        plural = 's'
        missing = 2
        arguments = "'slack' and 'event'"
    else:
        plural = ''
        missing = 1
        arguments = "'event'"

    return (f"{name}{sig} missing {missing} required positional "
            f"argument{plural}: {arguments}")
```
and
```python
if num_params < 2:
    raise TypeError(self._format_parameter_error_message(
        fn.__name__, sig, num_params))
```
[![Boy, that escalated quickly.](https://media.giphy.com/media/ToMjGpjpXMFPshSYGLm/giphy.gif)](https://media.giphy.com/media/ToMjGpjpXMFPshSYGLm/giphy.gif)

At least now it **actually** does what Python does.
```python
>>> from layabout import Layabout
>>> layabout = Layabout()
>>> @layabout.handle('hello')
>>> def invalid():
...     pass
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tucker/Projects/layabout/layabout.py", line 120, in decorator
    fn.__name__, sig, num_params))
TypeError: invalid() missing 2 required positional arguments: 'slack' and 'event'
>>> @layabout.handle('hello')
... def invalid(slack):
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tucker/Projects/layabout/layabout.py", line 120, in decorator
    fn.__name__, sig, num_params))
TypeError: invalid(slack) missing 1 required positional argument: 'event'
>>>
```